redash/handlers/dashboards.py

```diff
@@ -49,18 +77,45 @@ def get(self, dashboard_slug=None):
             response['public_url'] = url_for('redash.public_dashboard', token=api_key.api_key, org_slug=self.current_org.slug, _external=True)
             response['api_key'] = api_key.api_key
 
+        last_change = models.Change.get_latest(object_id=dashboard.id, object_type=models.Dashboard.__name__)
+        if last_change:
+            response['latest_version'] = last_change.id
+
         return response
 
     @require_permission('edit_dashboard')
     def post(self, dashboard_slug):
         dashboard_properties = request.get_json(force=True)
         # TODO: either convert all requests to use slugs or ids
         dashboard = models.Dashboard.get_by_id_and_org(dashboard_slug, self.current_org)
+
+        # check access permissions
+        if self.current_user.id != dashboard.user.id:
+            if not self.current_user.has_access(
+                    access_type=models.AccessPermission.ACCESS_TYPE_MODIFY,
+                    object_id=dashboard.id,
+                    object_type=models.Dashboard.__name__):
+                abort(403)
+
+        # Optimistic locking: figure out which user made the last
+        # change to this dashboard, and bail out if necessary
+        last_change = models.Change.get_latest(object_id=dashboard.id, object_type=models.Dashboard.__name__)
+        if last_change and 'latest_version' in dashboard_properties:
+            if last_change.id > dashboard_properties['latest_version']:
+                abort(409) # HTTP 'Conflict' status code
```

Comment:
This is *very* optimistic locking :-) As it hopes that no one else will save anything in the time that passes from this check until line 110. While usually it's OK, it's still a risk.

The optimistic locking I was referring to is what [Charles suggested in this comment](https://github.com/getredash/redash/issues/876#issuecomment-223488522):

```python
class BaseVersionedModel(BaseModel):
    version = IntegerField(default=0)

    def save_optimistic(self):
        if not self.version:
            return self.save()  # Since this is an `INSERT`, just call regular save method.

        # Update any data that has changed and bump the version counter.
        field_data = dict(self._data)
        current_version = field_data.pop('version', 0)
        field_data = self._prune_fields(field_data, self.dirty_fields)
        if not field_data:
            raise ValueError('No changes have been made.')

        ModelClass = type(self)
        field_data['version'] = ModelClass.version + 1  # Atomic increment

        query = ModelClass.update(**field_data).where(
            (ModelClass.version == current_version) &
            (ModelClass.id == self.id))

        nrows = query.execute()
        if nrows == 0:
            # It looks like another process has updated the version number.
            raise ConflictDetectedException()  # Raise exception? Return False?
        else:
            self.version += 1  # Update in-memory version number.
            return nrows
```

Note the use of update + where. It might be needed some changes, to use our callbacks for pre/post save.